examples/nlp/text_generation_gpt.py

```diff
@@ -0,0 +1,418 @@
+
+This example combines concepts from [Text generation with a miniature GPT](https://keras.io/examples/generative/text_generation_with_miniature_gpt/) 
+with KerasNLP abstractions. We will demonstrate how KerasNLP tokenization, model, metrics, and
+inference libraries can be used to simplify the development process. 
+
+The main KerasNLP features shown in this example are:
+
+- [WordPieceTokenizer](https://keras.io/api/keras_nlp/tokenizers/word_piece_tokenizer/):
+An efficient sub-word tokenizer used by BERT and other models.
+- [TokenAndPositionEmbedding](https://keras.io/api/keras_nlp/layers/token_and_position_embedding/): 
+A layer that learns an embedding based on the token and it's position in the
+input.
+- [TransformerDecoder](https://keras.io/api/keras_nlp/layers/transformer_decoder/): A
+decoder transformer layer with causal masking, as described in the transformers paper
+[Attention Is All You Need](https://arxiv.org/abs/1706.03762?context=cs).
+- [Text Generation Utilities](https://github.com/keras-team/keras-nlp/blob/v0.3.0/keras_nlp/utils/text_generation.py): 
+Utilities for different algorithms for text generation, including Beam Search,
+Top-K and Top-P.
+"""
+
+"""
+# Mini-GPT Text Generation with KerasNLP transformers
+
+This example requires KerasNLP. You can install it via the following command: `pip
+install keras-nlp`
+"""
+
+"""
+## Imports
+"""
+
+import os
+import keras_nlp
+import tensorflow as tf
+from tensorflow import keras
+
+"""
+## Settings / Hyperparameters
+"""
+
+# Data
+SEQ_LEN = 128
+MIN_TRAINING_SEQ_LEN = 450
+
+# Inference
+NUM_TOKENS_TO_GENERATE = 80
+
+# Training
+LEARNING_RATE = 5e-4
+EPOCHS = 12
+NUM_TRAINING_BATCHES = 1000
+
+# Model
+BATCH_SIZE = 64
+EMBED_DIM = 256
+FEED_FORWARD_DIM = 256
+NUM_HEADS = 3
+NUM_LAYERS = 2
+VOCAB_SIZE = 5000  # Limits parameters in model.
+
+"""
+## Load Tokenizer
+
+Here we download the vocabulary data and initialize
+`keras_nlp.tokenizers.WordPieceTokenizer`. WordPieceTokenizer is an efficient
+implementation of the WordPiece algorithm used by BERT and other models. It will strip,
+lower-case and do other un-reversable preprocessing operations.
+
+The vocabulary data we download is a list of sub-words sorted in descending order by
+frequency. We want to limit the vocabulary as much as possible, as we will see later on
+that it has a large affect on the number of model parameters. We also don't want to take
+too few vocabulary, or there would be too many OOV sub-words. 
+
+We strip the first 1000 vocabulary that are all of the form `"[unusedXX]"`, since they
+are all not useful for this text corpus and unneccesarily increases vocabulary size. Then
+we take the `VOCAB_SIZE` most frequent vocabularies after that. There are also some
+reserved tokens, so we want to make sure that `"[PAD]"` is at index 0 and `"[UNK]"` is in
+the vocabulary, since it is used for padding and unknown words respectively.
+"""
+
+# Download vocabulary data.
+vocab_file = keras.utils.get_file(
```

Comment:
should we use the word piece vocab learner utility here?

Reply:
edited